Json parsing exceptions should not cause 500 errors

cbuescher opened the pull request:
Currently we wrap JsonEOFException from advancing the json parser into our own XContentEOFException, but this has the drawback that is results in 500 errors on the client side. Instead this should be 400 errors. We should throw XContentParseException instead like we e.g. already to in the "nextFieldName" method or elsewhere in JsonXContentParser.

Closes #111542

cbuescher commented:
tagging @elastic/es-core-infra for awareness and you own general xcontent issues, but we found this while looking into 500s from the "_search" endpoint, so also tagging search foundations.

elasticsearchmachine commented:
Hi @cbuescher, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

cbuescher commented:
I ended up removing the whole XContentEOFException, the only place we differentiate based on the cause of an XContentParseException seems to be the warning in the BulkRequestParser that is going to be removed anyway.

cbuescher commented on server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:
```diff
@@ -421,12 +421,22 @@ private static void checkBulkActionIsProperlyClosed(XContentParser parser) throw
+        } catch (XContentParseException e) {
+            Throwable cause = e.getCause();
+            if (cause != null) {
+                String exName = cause.getClass().getName();
```
This workaround was necessary because we don't seem to have JsonEOFException on the classpath here. The class name is readable at runtime though it seems.

rjernst replied:
You should not assume JsonEOFException exists, jackson is intentionally hidden, it is an implementation detail of xcontent.

dnhatn commented:
We can also throw IllegalStateException while parsing XContent.

```
java.lang.IllegalStateException: Can't get text on a START_ARRAY at 1:30
	at org.elasticsearch.xcontent.impl@8.16.0/org.elasticsearch.xcontent.provider.json.JsonXContentParser.throwOnNoText(JsonXContentParser.java:113)
	at org.elasticsearch.xcontent.impl@8.16.0/org.elasticsearch.xcontent.provider.json.JsonXContentParser.text(JsonXContentParser.java:107)
	at org.elasticsearch.xcontent.impl@8.16.0/org.elasticsearch.xcontent.provider.json.JsonXContentParser.objectText(JsonXContentParser.java:139)
	at org.elasticsearch.xcontent@8.16.0/org.elasticsearch.xcontent.FilterXContentParser.objectText(FilterXContentParser.java:114)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.MatchQueryBuilder.fromXContent(MatchQueryBuilder.java:549)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.SearchModule.lambda$registerQuery$23(SearchModule.java:1279)
	at org.elasticsearch.xcontent@8.16.0/org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:148)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder$1.namedObject(AbstractQueryBuilder.java:402)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.parseInnerQueryBuilder(AbstractQueryBuilder.java:436)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.parseTopLevelQuery(AbstractQueryBuilder.java:410)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SearchSourceBuilder.parseXContent(SearchSourceBuilder.java:1417)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SearchSourceBuilder.parseXContent(SearchSourceBuilder.java:1298)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.rest.action.search.RestSearchAction.parseSearchRequest(RestSearchAction.java:181)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.rest.action.search.RestSearchAction.lambda$prepareRequest$1(RestSearchAction.java:120)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.rest.RestRequest.withContentOrSourceParamParserOrNull(RestRequest.java:544)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.rest.action.search.RestSearchAction.prepareRequest(RestSearchAction.java:119)
```

dnhatn commented:
Another error

```
com.fasterxml.jackson.core.JsonParseException: Illegal unquoted character ((CTRL-CHAR, code 10)): has to be escaped using backslash to be included in string value
 at [Source: (byte[])\"
{
  \"query\": {
    \"query_string\": {
      \"query\": \"KAFFE TRINA JUMPR Celestial Blue
xtra Small\"
    }
  }
}\"; line: 5, column: 50]
\tat com.fasterxml.jackson.core@2.15.0/com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2477)
\tat com.fasterxml.jackson.core@2.15.0/com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:750)
\tat com.fasterxml.jackson.core@2.15.0/com.fasterxml.jackson.core.base.ParserBase._throwUnquotedSpace(ParserBase.java:1374)
\tat com.fasterxml.jackson.core@2.15.0/com.fasterxml.jackson.core.json.UTF8StreamJsonParser._finishString2(UTF8StreamJsonParser.java:2631)
\tat com.fasterxml.jackson.core@2.15.0/com.fasterxml.jackson.core.json.UTF8StreamJsonParser._finishAndReturnString(UTF8StreamJsonParser.java:2560)
\tat com.fasterxml.jackson.core@2.15.0/com.fasterxml.jackson.core.json.UTF8StreamJsonParser.getText(UTF8StreamJsonParser.java:335)
\tat org.elasticsearch.xcontent.impl@8.16.0/org.elasticsearch.xcontent.provider.json.JsonXContentParser.text(JsonXContentParser.java:109)
\tat org.elasticsearch.xcontent@8.16.0/org.elasticsearch.xcontent.FilterXContentParser.text(FilterXContentParser.java:94)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.QueryStringQueryBuilder.fromXContent(QueryStringQueryBuilder.java:690)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.search.SearchModule.lambda$registerQuery$23(SearchModule.java:1279)
\tat org.elasticsearch.xcontent@8.16.0/org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:148)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder$1.namedObject(AbstractQueryBuilder.java:402)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.parseInnerQueryBuilder(AbstractQueryBuilder.java:436)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.parseTopLevelQuery(AbstractQueryBuilder.java:410)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SearchSourceBuilder.parseXContent(SearchSourceBuilder.java:1417)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SearchSourceBuilder.parseXContent(SearchSourceBuilder.java:1298)
\tat org.elasticsearch.server@8.16.0/org.elasticsearch.rest.action.search.RestSearchAction.parseSearchRequest(RestSearchAction.java:181)
```

elasticsearchmachine commented:
Hi @cbuescher, I've created a changelog YAML for you.

cbuescher committed:
Json parsing exceptions should not cause 500 errors

Currently we wrap JsonEOFException from advancing the json parser into our own
XContentEOFException, but this has the drawback that is results in 500 errors on
the client side. Instead this should be 400 errors. We should throw
XContentParseException instead like we e.g. already to in the "nextFieldName"
method or elsewhere in JsonXContentParser.

Closes #111542

cbuescher commented:
@rjernst thanks for the review, I changed the base class of XContentEOFException as suggested. This should be ready for another review now.

cbuescher committed:
Spotless and changelog

cbuescher committed:
moa checkstyle

cbuescher committed:
Fix test

cbuescher committed:
iter

cbuescher committed:
another yaml

rjernst reviewed (approved):
LGTM

cbuescher commented:
@rjernst thanks a lot, merging

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.14  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.15  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 111548`
